posts/2020/03/pytorch-tensoriterator-internals.md

```diff
@@ -0,0 +1,267 @@
+
+## Limitations of TH iterators
+
+Apart from the obvious complication that arises due to maintaining a codebase that is so dependent
+on such insanely complex macro expansions, TH iterators have some fundamental shortcomings. For
+one thing, they cannot be used for writing iterators in a device independent manner - you will
+need separate iterators for CPU and CUDA. Also, parallelization does not happen implicitly
+inside the iterator, you need to write the parallel looping logic yourself. Moreover, at a deeper
+level `TH` iterators do not collapse the dimensions of the tensor (as we'll see later in this
+post) therefore leading to looping that might not be as cache-optimized as possible.
+
+These limitations led to the creation of `TensorIterator`, which is used by the
+`ATen` tensor implementation for overcoming some of the shortcomings of the previous `TH`
+iterators.
+
+# Basics of TensorIterator
+
+A `TensorIterator` can be created using the default constructor. You must then add the tensors
+that you want as inputs or outputs. A good example can be found from the `TensorIterator::binary_op()`
+[method](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.cpp#L652) that
+allows you to create `TensorIterator` objects for performing point-wise binary operations
+between two tensors. The important parts look like so:
+
+``` cpp
+auto iter = TensorIterator();
+
+iter.add_output(out);
+iter.add_input(a);
+iter.add_input(b);
+
+iter.build();
+```
+As you can see, you add a tensor called `out` as the output tensors and `a` and `b` as the
+input tensors. Calling `build` is then mandatory for creating the object and letting
+the class perform other optimizations like collapsing dimensions.
+
+# Performing iterations
+
+Broadly, iterations using `TensorIterator` can be classified as point-wise iterations
+or reduction iterations. This plays a fundamental role in how iterations using `TensorIterator`
+are parallelized - point-wise iterations can be freely parallelized along any dimension
+and grain size while reduction operations have to be either parallelized along dimensions
+that you're not iterating over or by performing bisect and reduce operations along the
+dimension being iterated. Parallelization can also happen using vectorized operations.
+
+## Iteration details
+
+The simplest iteration operation can be performed using the 
+[`for_each`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.cpp#L525) 
+function. This function has two overloads: one takes a function object which iterates over a
+single dimension (`loop_t`); the other takes a function object which iterates over two
+dimensions simultaneously (`loop2d_t`). Find their definitions [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.h#L166). The former can iterate over a loop
```

Comment:
It seems like it is only true for CPU. For CUDA it looks like TensorIterator primarily just stores the pointers. In order to process CUDA tensors, you have to write your own loop which is not passed to the `for_each` procedure. A good example of this is `IndexKernel.cu`.

Reply:
Although I state at the top that `TensorIterator` is device agnostic, the post has been written keeping CPU in mind. I will specify this at the top of the post.